Modules/StudyProgramme/classes/class.ilObjStudyProgrammeSettingsGUI.php

```diff
@@ -191,50 +213,79 @@ protected function buildModalHeading($label, $current_node) {
 	const PROP_POINTS = "points";
 	const PROP_STATUS = "status";
 
-	protected function buildForm() {
-		$form = new ilAsyncPropertyFormGUI();
+	protected function buildForm(\ilObjStudyProgramme $prg, string $submit_action) : ILIAS\UI\Component\Input\Container\Form\Standard {
+		$ff = $this->input_factory->field();
+		$tf = $this->trafo_factory;
+		$txt = function($id) { return $this->lng->txt($id); };
+		$sp_types = ilStudyProgrammeType::getAllTypesArray();
+		$status_options = self::getStatusOptions();
+		return $this->input_factory->container()->form()->standard(
+			$submit_action,
+			[
+				$ff->section(
+					[
+						self::PROP_TITLE =>
+							$ff->text($txt("title"))
+								->withValue($prg->getTitle())
+								->withRequired(true),
+						self::PROP_DESC =>
+							$ff->text($txt("description")) // TODO: change this to textarea once available
+								->withValue($prg->getDescription())
+					],
+					$txt("prg_edit"),
+					""
+				),
+				$ff->section(
+					[
+						self::PROP_TYPE =>
+							$ff->select($txt("type"), $sp_types)
+								->withValue((string)$prg->getSubtypeId())
+								->withAdditionalTransformation($tf->custom(function($v) {
+									if ($v == "") {
+										return 0;
+									}
+									return $v;
+								}))
+					],
+					$txt("prg_type"),
+					""
+				),
+				$ff->section(
+					[
+						self::PROP_POINTS =>
+							$ff->numeric($txt("prg_points"))
+								->withValue((string)$prg->getPoints()),
+						self::PROP_STATUS =>
+							$ff->select($txt("prg_status"), $status_options)
+								->withValue((string)$prg->getStatus())
+					],
+					$txt("prg_assessment"),
+					""
+				)
+			]
+		)
+		->withAdditionalTransformation($tf->custom(function($values) {
+			// values now contains the results of the single sections,
+			// i.e. a list of arrays that each contains keys according
+			// to the section they originated from.
+			return call_user_func_array("array_merge", $values);
+		}));
+	}
```

Comment:
I do not really like that much nesting in terms of shuffling arrays each other. However might look at this, note, that you might also generate your inputs like so:
```
$input_a = $ff->numeric($txt("prg_points"))->...
$input_b = $ff->select($txt("prg_status"), ....
```

an then pack them subsequentially:
`$ff->section([$input_a ,$input_b],$txt("prg_assessment"),"")`

and so on. I think there is no good or bad way here, it's just a matter of personal taste.

Reply:
Yes, I partially agree. In the style I used, without temporary variables, one directly finds the structure of the form as it is shown on the GUI without following references. In general I like that, but this might also get out of hand at some point. ATM I think I would start to use methods for the single sections once the form gets bigger then, say, one page one my screen. I would use methods (and not temporary variables) because this would allow me to have a reading flow from general-form-layout to specific sections and inputs, while I would need to define the specifics (sections, inputs, etc.) before the general structure if I use variables.